File: templates/ruflet_flutter_template/apple_packages/ruflet_apple/Sources/RufletEngine/Controls/pagelet.swift
import RufletProtocol
import SwiftUI

/// Apple-native port of Flet's `pagelet.dart`.
@MainActor
public struct PageletControl: View {
  @ObservedObject public var control: RufletControl
  @State private var drawerPresented = false
  @State private var endDrawerPresented = false
  @State private var invokeToken: UUID?

  public init(control: RufletControl) {
    self.control = control
  }

  public var body: some View {
    LayoutControl(control: control) {
      GeometryReader { proxy in
        if proxy.size.height.isInfinite && control.number("height") == nil {
          ErrorControl(
            "Error displaying Pagelet: height is unbounded.",
            description: "Either set a fixed height or nest Pagelet inside expanded control or control with a fixed height.")
        } else if control.child("content") == nil {
          ErrorControl("Pagelet.content must be provided and visible")
        } else {
          scaffold
        }
      }
    }
    .onAppear(perform: mount)
    .onDisappear(perform: unmount)
  }

  private var scaffold: some View {
    ZStack {
      parseColor(control.string("bgcolor")) ?? Color.rufletSystemBackground
      VStack(spacing: 0) {
        if let appBar = control.child("appbar") {
          pageletAppBar(appBar)
        }
        control.buildWidget("content")!
          .frame(maxWidth: .infinity, maxHeight: .infinity)
        if let bottom = control.child("navigation_bar") ?? control.child("bottom_appbar") {
          ControlWidget(control: bottom)
        }
      }
      if let sheet = control.child("bottom_sheet") {
        ControlWidget(control: sheet)
      }
      if let floating = control.child("floating_action_button") {
        ControlWidget(control: floating)
          .padding(16)
          .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: floatingAlignment)
      }
      drawerLayers
    }
    .clipped()
  }

  @ViewBuilder
  private func pageletAppBar(_ appBar: RufletControl) -> some View {
    if pageletDesign == .cupertino || appBar.type == "CupertinoAppBar" {
      RufletAppleAppBar(control: appBar, kind: .cupertino)
    } else {
      AppBarControl(control: appBar)
    }
  }

  private var pageletDesign: RufletPageDesign {
    rufletPageDesign(
      adaptive: control.boolean("adaptive", default: false),
      platform: nil,
      defaultPlatform: rufletDefaultTargetPlatform)
  }

  @ViewBuilder
  private var drawerLayers: some View {
    if let drawer = control.child("drawer"), drawerPresented {
      RufletPageletDrawer(edge: .leading) {
        ControlWidget(control: drawer)
      } onDismiss: {
        drawerPresented = false
        control.backend.triggerControlEvent(controlID: drawer.id, name: "dismiss", data: .null)
      }
    }
    if let drawer = control.child("end_drawer"), endDrawerPresented {
      RufletPageletDrawer(edge: .trailing) {
        ControlWidget(control: drawer)
      } onDismiss: {
        endDrawerPresented = false
        control.backend.triggerControlEvent(controlID: drawer.id, name: "dismiss", data: .null)
      }
    }
  }

  private func mount() {
    guard invokeToken == nil else { return }
    invokeToken = control.addInvokeMethodListener { name, _ in
      switch name {
      case "show_drawer": drawerPresented = control.child("drawer") != nil
      case "close_drawer": drawerPresented = false
      case "show_end_drawer": endDrawerPresented = control.child("end_drawer") != nil
      case "close_end_drawer": endDrawerPresented = false
      default: throw RufletPageletError.unknownMethod(name)
      }
      return .null
    }
  }

  private func unmount() {
    if let invokeToken { control.removeInvokeMethodListener(invokeToken) }
    invokeToken = nil
  }

  private var floatingAlignment: Alignment {
    switch control.string("floating_action_button_location")?.lowercased() {
    case let value? where value.contains("center"): .bottom
    case let value? where value.contains("start"): .bottomLeading
    default: .bottomTrailing
    }
  }
}

private enum RufletPageletError: LocalizedError {
  case unknownMethod(String)
  var errorDescription: String? {
    switch self { case .unknownMethod(let name): "Unknown Pagelet method: \(name)" }
  }
}

private struct RufletPageletDrawer<Content: View>: View {
  let edge: Edge
  @ViewBuilder let content: () -> Content
  let onDismiss: () -> Void

  var body: some View {
    ZStack(alignment: edge == .leading ? .leading : .trailing) {
      Color.black.opacity(0.28)
        .ignoresSafeArea()
        .onTapGesture(perform: onDismiss)
      content()
        .frame(maxWidth: 360, maxHeight: .infinity)
        .background(Color.rufletSystemBackground)
        .transition(.move(edge: edge))
    }
    .accessibilityAddTraits(.isModal)
  }
}

File: templates/ruflet_flutter_template/ruflet_packages/ruflet/lib/src/controls/pagelet.dart
import 'package:flutter/widgets.dart';

import '../extensions/control.dart';
import '../models/control.dart';
import '../models/control_type.dart';
import '../models/page_design.dart';
import '../utils/colors.dart';
import '../utils/numbers.dart';
import '../widgets/error.dart';
import '../widgets/ruflet_store_mixin.dart';
import '../widgets/platform_page_scaffold.dart';
import '../widgets/platform_control_renderer.dart';
import 'adaptive_app_bar.dart';
import 'base_controls.dart';
import 'control_widget.dart';

class PageletControl extends StatefulWidget {
  final Control control;

  PageletControl({Key? key, required this.control})
      : super(key: key ?? ValueKey("control_${control.id}"));

  @override
  State<PageletControl> createState() => _PageletControlState();
}

class _PageletControlState extends State<PageletControl> with RufletStoreMixin {
  final _scaffoldController = PlatformPageScaffoldController();

  @override
  void initState() {
    super.initState();
    widget.control.addInvokeMethodListener(_invokeMethod);
  }

  @override
  void dispose() {
    widget.control.removeInvokeMethodListener(_invokeMethod);
    super.dispose();
  }

  Future<dynamic> _invokeMethod(String name, dynamic args) async {
    debugPrint("Pagelet.$name($args)");
    switch (name) {
      case "show_drawer":
        _scaffoldController.showDrawer();
        break;
      case "close_drawer":
        _scaffoldController.closeDrawer();
        break;
      case "show_end_drawer":
        _scaffoldController.showEndDrawer();
        break;
      case "close_end_drawer":
        _scaffoldController.closeEndDrawer();
        break;
      default:
        throw Exception("Unknown Pagelet method: $name");
    }
  }

  @override
  Widget build(BuildContext context) {
    return withPagePlatform(_buildForPlatform);
  }

  Widget _buildForPlatform(BuildContext context, TargetPlatform platform) {
    debugPrint("Pagelet build: ${widget.control.id}");

    var appBar = widget.control.child("appbar");
    var content = widget.control.buildWidget("content");
    var navigationBar = widget.control.buildWidget("navigation_bar");
    var bottomAppBar = widget.control.buildWidget("bottom_appbar");
    var bottomSheet = widget.control.buildWidget("bottom_sheet");
    var drawer = widget.control.child("drawer");
    var endDrawer = widget.control.child("end_drawer");
    var hasDrawer = drawer != null || endDrawer != null;
    var fab = widget.control.buildWidget("floating_action_button");

    if (content == null) {
      return const ErrorControl("Pagelet.content must be provided and visible");
    }

    var widgetsDesign = usesCupertinoControls(platform)
        ? PageDesign.cupertino
        : PageDesign.material;

    var bnb = navigationBar ?? bottomAppBar;

    void dismissDrawer(int id) {
      widget.control.backend.triggerControlEventById(id, "dismiss");
    }

    var bar = appBar != null
        ? appBar.canonicalType == "AppBar"
            ? AdaptiveAppBarControl(control: appBar)
            : null
        : null;

    Widget scaffold = PlatformPageScaffold(
        controller: _scaffoldController,
        design: widgetsDesign,
        backgroundColor: widget.control.getColor("bgcolor", context),
        appBar: bar,
        drawer: drawer != null ? ControlWidget(control: drawer) : null,
        onDrawerChanged: (opened) {
          if (drawer != null && !opened) {
            dismissDrawer(drawer.id);
          }
        },
        endDrawer: endDrawer != null ? ControlWidget(control: endDrawer) : null,
        onEndDrawerChanged: (opened) {
          if (endDrawer != null && !opened) {
            dismissDrawer(endDrawer.id);
          }
        },
        body: content,
        bottomNavigationBar: bnb,
        bottomSheet: bottomSheet,
        floatingActionButton: fab,
        floatingActionButtonLocation:
            widget.control.get("floating_action_button_location"));

    if (hasDrawer) {
      // Clip to page bounds so the drawer animation stays hidden outside the pagelet.
      scaffold = ClipRect(child: scaffold);
    }

    final scaffoldWithBoundsCheck = scaffold;

    scaffold = LayoutBuilder(
        builder: (BuildContext context, BoxConstraints constraints) {
      debugPrint("Pagelet constraints.maxWidth: ${constraints.maxWidth}");
      debugPrint("Pagelet constraints.maxHeight: ${constraints.maxHeight}");

      if (constraints.maxHeight == double.infinity &&
          widget.control.getDouble("height") == null) {
        return const ErrorControl(
            "Error displaying Pagelet: height is unbounded.",
            description:
                "Either set a fixed \"height\" or nest Pagelet inside expanded control or control with a fixed height.");
      }

      return scaffoldWithBoundsCheck;
    });

    return LayoutControl(control: widget.control, child: scaffold);
  }
}
